once/__init__.py

```diff
@@ -25,110 +21,167 @@ def _is_method(func: collections.abc.Callable):
     if inspect.isfunction(func):
         return _WrappedFunctionType.SYNC_FUNCTION
-    return _WrappedFunctionType.UNSUPPORTED
+    raise SyntaxError(f"Unable to determine function type for {repr(original_func)}")
+
 
+class _ExecutionState(enum.Enum):
+    UNCALLED = 0
+    WAITING = 1
+    COMPLETED = 2
 
-class _OnceBase(abc.ABC):
+
+class _OnceBase:
     """Abstract Base Class for once function decorators."""
 
-    def __init__(self, func: collections.abc.Callable) -> None:
-        functools.update_wrapper(self, func)
-        self.func = self._inspect_function(func)
-        self.called = False
+    def __init__(self, fn_type: _WrappedFunctionType) -> None:
+        self.call_state = _ExecutionState.UNCALLED
         self.return_value: typing.Any = None
-        self.fn_type = _wrapped_function_type(self.func)
-        if self.fn_type == _WrappedFunctionType.UNSUPPORTED:
-            raise SyntaxError(f"Unable to wrap a {type(func)}")
-        if self.fn_type == _WrappedFunctionType.ASYNC_FUNCTION:
+        self.fn_type = fn_type
+        if (
+            self.fn_type == _WrappedFunctionType.ASYNC_FUNCTION
+            or self.fn_type == _WrappedFunctionType.ASYNC_GENERATOR
+        ):
             self.async_lock = asyncio.Lock()
         else:
-            self.lock = _new_lock()
-
-    @abc.abstractmethod
-    def _inspect_function(self, func: collections.abc.Callable) -> collections.abc.Callable:
-        """Inspect the passed-in function to ensure it can be wrapped.
+            self.lock = threading.Lock()
 
-        This function should raise a SyntaxError if the passed-in function is
-        not suitable.
+    def _callable(self, func: collections.abc.Callable):
+        """Generate a wrapped function appropriate to the function type.
 
-        It should return the function which should be executed once.
+        This wrapped function will call the correct _execute_call_once function.
         """
+        if self.fn_type == _WrappedFunctionType.ASYNC_GENERATOR:
+
+            async def wrapped(*args, **kwargs):
+                next_value = None
+                iterator = self._execute_call_once_async_iter(func, *args, **kwargs)
+                while True:
+                    try:
+                        next_value = yield await iterator.asend(next_value)
+                    except StopAsyncIteration:
+                        return
+
+        elif self.fn_type == _WrappedFunctionType.ASYNC_FUNCTION:
+
+            async def wrapped(*args, **kwargs):
+                return await self._execute_call_once_async(func, *args, **kwargs)
+
+        elif self.fn_type == _WrappedFunctionType.SYNC_FUNCTION:
+
+            def wrapped(*args, **kwargs):
+                return self._execute_call_once_sync(func, *args, **kwargs)
+
+        else:
+            if self.fn_type != _WrappedFunctionType.SYNC_GENERATOR:
+                print(self.fn_type)
+            assert self.fn_type == _WrappedFunctionType.SYNC_GENERATOR
```

Comment:
I'm curious why you have this pattern vs something like 

```python3
elif self.fn_type == _WrappedFunctionType.SYNC_GENERATOR:
  ...
else:
  raise some_exception
```

Reply:
Good question. Generally, I feel like raising exceptions is preferable to assert. However, in this library, we are using assert liberally **ONLY** to express our assumptions throughout the library. All of our code will function identically without these assert statements, but this way it is clear what the assumption is, and when we run our unit tests, we also get to ensure that our assumptions always hold. Crucially, in a case like this, the internal state is not dependent on outside input.